internal/tasks/account_deletion.go

```diff
@@ -0,0 +1,213 @@
+    WHERE a.name = $1 AND parent_digest IS NULL
+	`)
+	deleteAccountReposQuery                   = `DELETE FROM repos WHERE account_name = $1`
+	deleteAccountCountBlobsQuery              = `SELECT COUNT(id) FROM blobs WHERE account_name = $1`
+	deleteAccountScheduleBlobSweepQuery       = `UPDATE accounts SET next_blob_sweep_at = $2 WHERE name = $1`
+	deleteAccountMarkAllBlobsForDeletionQuery = `UPDATE blobs SET can_be_deleted_at = $2 WHERE account_name = $1`
+)
+
+type deleteAccountRemainingManifest struct {
+	RepositoryName string
+	Digest         string
+}
+
+func (j *Janitor) DeleteAccount(ctx context.Context, account models.Account, actx keppel.AuditContext) error {
+	// can only delete account when user has deleted all manifests from it
+	var nextManifests []deleteAccountRemainingManifest
+	err := sqlext.ForeachRow(j.db, deleteAccountFindManifestsQuery, []any{account.Name},
+		func(rows *sql.Rows) error {
+			var m deleteAccountRemainingManifest
+			err := rows.Scan(&m.RepositoryName, &m.Digest)
+			nextManifests = append(nextManifests, m)
+			return err
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(nextManifests) > 0 {
+		// deletion was not completed yet -> check if we need to do something on our side
+		logg.Info("cleaning up account %q: need to delete %d manifests in this cycle", account.Name, len(nextManifests))
+		for _, rm := range nextManifests {
+			parsedDigest, err := digest.Parse(rm.Digest)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: could not parse digest: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+			repo, err := keppel.FindRepository(j.db, rm.RepositoryName, account.Name)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: could not find repository in DB: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+			err = j.processor().DeleteManifest(ctx, account.Reduced(), *repo, parsedDigest, actx)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+		}
+	}
+
+	// delete all repos (and therefore, all blob mounts), so that blob sweeping can immediately take place
+	_, err = j.db.Exec(deleteAccountReposQuery, account.Name)
+	if err != nil {
+		return err
+	}
+
+	// can only delete account when all blobs have been deleted
+	blobCount, err := j.db.SelectInt(deleteAccountCountBlobsQuery, account.Name)
+	if err != nil {
+		return err
+	}
+	if blobCount > 0 {
+		// make sure that blob sweep runs immediately
+		// TODO: how to prevent resetting time stamp if already set?
+		_, err := j.db.Exec(deleteAccountMarkAllBlobsForDeletionQuery, account.Name, j.timeNow())
+		if err != nil {
+			return err
+		}
+		_, err = j.db.Exec(deleteAccountScheduleBlobSweepQuery, account.Name, j.timeNow())
+		if err != nil {
+			return err
+		}
+		logg.Info("cleaning up managed account %q: waiting for %d blobs to be deleted", account.Name, blobCount)
```

Comment:
When deferring into a different job, you should update this job's clock field (`next_deletion_attempt_at`) by a little bit (e.g. 1 minute) to avoid this job getting hammered while it cannot make any progress.